src/runtime/program/vote/execute.zig

```diff
@@ -138,10 +174,287 @@ fn intializeAccount(
+
+    switch (vote_authorize) {
+        .voter => {
+            const current_epoch = clock.epoch;
+            const target_epoch = std.math.add(u64, current_epoch, 1) catch {
+                return InstructionError.InvalidAccountData;
+            };
+
+            try checkAuthority(
+                allocator,
+                ic,
+                &vote_state,
+                signers,
+                current_epoch,
+            );
+
+            vote_state.setNewAuthorizedVoter(
+                authorized,
+                target_epoch,
+            ) catch |err| {
+                switch (err) {
+                    VoteError.TooSoonToReauthorize => {
+                        ic.tc.custom_error = @intFromError(err);
+                        return InstructionError.Custom;
+                    },
+                    else => {
+                        return InstructionError.InvalidAccountData;
+                    },
+                }
+            };
+        },
+        .withdrawer => {
+            // current authorized withdrawer must say "yay".
+            const authorized_withdrawer_signer = if (signers) |signers_|
+                try verifyAuthorizedSigner(vote_state.authorized_withdrawer, signers_)
+            else
+                ic.info.isPubkeySigner(vote_state.authorized_withdrawer);
+
+            if (!authorized_withdrawer_signer) {
+                return InstructionError.MissingRequiredSignature;
+            }
+            vote_state.authorized_withdrawer = authorized;
+        },
+    }
+    try vote_account.serializeIntoAccountData(VoteStateVersions{ .current = vote_state });
+}
+
+/// Analogous to
+/// https://github.com/anza-xyz/agave/blob/49fb51295c1062b6b09e585b2fe0a4676c33d3d4/programs/vote/src/vote_state/mod.rs#L691-L692
+/// https://github.com/anza-xyz/agave/blob/49fb51295c1062b6b09e585b2fe0a4676c33d3d4/programs/vote/src/vote_state/mod.rs#L701-L707
+/// https://github.com/anza-xyz/solana-sdk/blob/4e30766b8d327f0191df6490e48d9ef521956495/vote-interface/src/state/mod.rs#L873
+///
+/// Check if the current authorized withdrawer or epoch authorized voter is a signer.
+fn checkAuthority(
+    allocator: std.mem.Allocator,
+    ic: *InstructionContext,
+    vote_state: *VoteState,
+    signers: ?std.AutoHashMap(Pubkey, void),
+    current_epoch: Epoch,
+) InstructionError!void {
+    const authorized_withdrawer_signer = if (signers) |signers_|
+        try verifyAuthorizedSigner(vote_state.authorized_withdrawer, signers_)
+    else
+        ic.info.isPubkeySigner(vote_state.authorized_withdrawer);
+
+    // current authorized withdrawer or epoch authorized voter must say "yay"
+    if (!authorized_withdrawer_signer) {
+        const epoch_authorized_voter = try vote_state.getAndUpdateAuthorizedVoter(
+            allocator,
+            current_epoch,
+        );
+        if (!ic.info.isPubkeySigner(epoch_authorized_voter)) {
+            return InstructionError.MissingRequiredSignature;
+        }
```

Comment:
validate methods shouldnt have state modification code in them imo - would be good to pull this out and refactor

Reply:
> validate methods shouldnt have state modification code in them imo

I agree.

While trying to pull it out, I ended up dropping the method altogether. This is because `vote_state.getAndUpdateAuthorizedVoter` which contains the state update is core to the check that is being performed but it should only be called if initial check is `authorized_withdrawer_signer`. Trying to pull this out would mean the check function would be redundant or such that it calls a passed in function (which contains the vote_state.getAndUpdateAuthorizedVoter call)


[694a8c6](https://github.com/Syndica/sig/pull/576/commits/694a8c64b1f56d8b9f4531069ea163b9d78be19d)